pyfr/readers/cgns.py

```diff
@@ -0,0 +1,423 @@
+
+        # Read Node
+        self.nodepts = self._read_nodepts(zone)
+
+        # Read bc
+        bc = self._read_bc(zone)
+
+        # Read elements
+        ns = cgns.nsections(zone)
+        self.elenodes = elenodes = {}
+        self.pents = pents = {}
+
+        # Construct elenodes and physical entity
+        newp = 0
+        for idx in range(ns):
+            elerng, elenode = self._read_element(zone, idx)
+
+            name = 'fluid'
+            for bcname, bcrng in bc.items():
+                if elerng[0] >= bcrng[0] and elerng[1] <= bcrng[1]:
+                    name = bcname
+
+            if name in pents:
+                pent = pents[name]
+            else:
+                pent = newp
+                pents[name] = pent
+                newp += 1
+
+            elenodes.update({(k, pent): v for k, v in elenode.items()})
+
+    def _read_nodepts(self, zone):
+        # Node read
+        nnode = zone['size'][0]
+        nodepts = np.zeros((3, nnode))
+
+        self._cgns.coord_read(zone, 'CoordinateX', nodepts[0])
+        self._cgns.coord_read(zone, 'CoordinateY', nodepts[1])
+        self._cgns.coord_read(zone, 'CoordinateZ', nodepts[2])
+
+        return nodepts
+
+    def _read_bc(self, zone):
+        nbc = self._cgns.nbocos(zone)
+        bc = {}
+
+        for idx_bc in range(nbc):
+            boco = self._cgns.boco_read(zone, idx_bc)
+            name = boco['name'].lower()
+            bc[name] = boco['range']
+
+        return bc
+
+    def _read_element(self, zone, idx):
+        s = self._cgns.section_read(zone, idx)
+
+        elerng = s['range']
+        conn = np.zeros(s['dim'], dtype=np.int32)
+        self._cgns.elements_read(s, conn)
+
+        try:
+            cgns_type = s['etype']
+            petype = self._cgns_map[cgns_type][0]
+        except KeyError as err:
+            raise KeyError('Unsupported elements: {}'.format(err))
+
+        elenode = {}
+
+        if petype == 'mixed':
+            i = 0
+            mele = defaultdict(list)
+
+            while i < s['dim']:
+                try:
+                    cgns_type = conn[i]
+                    petype, spts = self._cgns_map[cgns_type]
+                except KeyError as err:
+                    raise KeyError('Unsupported elements: {}'.format(err))
+
+                mele[cgns_type].append(conn[i + 1: i + 1 + spts])
+                i += 1 + spts
+
+            elenode = {k: np.array(v) for k, v in mele.items()}
+        else:
+            spts = self._cgns_map[cgns_type][1]
+            elenode[cgns_type] = conn.reshape(-1, spts)
+
+        return elerng, elenode
+
+
+class CGNSReader(BaseMeshConverter, BaseReader):
+    # Supported file types and extensions
+    name = 'cgns'
+    extn = ['.cgns']
+
+    # CGNS element types to PyFR type (petype) and node counts
+    _etype_map = CGNSZoneReader._cgns_map
+
+    # Number of nodes in the first-order representation an element
+    _petype_focount = {'line': 2, 'tri': 3, 'quad': 4,
```

Comment:
I think this can be moved into the BaseMeshReader class as both the Gmsh reader and the CGNS reader appear to need it.